research/minigo/dual_net.py

```diff
@@ -0,0 +1,209 @@
+    outputs = self.sess.run(
+        self.inference_output, feed_dict={self.inference_input: processed})
+    probabilities, value = outputs['policy_output'], outputs['value_output']
+    if use_random_symmetry:
+      probabilities = symmetries.invert_symmetries_pi(syms_used, probabilities)
+    return probabilities, value
+
+
+def get_inference_input():
+  """Set up placeholders for input features/labels.
+
+  Returns:
+    The features and output tensors that get passed into model_fn.
+  """
+  input_features = tf.placeholder(
+      tf.float32, [None, go.N, go.N, features.NEW_FEATURES_PLANES],
+      name='pos_tensor')
+
+  labels = {
+      'pi_tensor': tf.placeholder(tf.float32, [None, go.N * go.N + 1]),
+      'value_tensor': tf.placeholder(tf.float32, [None])
+  }
+
+  return input_features, labels
+
+
+def bootstrap(working_dir, params):
+  """Initialize a tf.Estimator run with random initial weights.
+
+  Args:
+    working_dir: The directory where tf.estimator will drop logs,
+      checkpoints, and so on
+    params: hyperparams of the model.
+  """
+  # Forge an initial checkpoint with the name that subsequent Estimator will
+  # expect to find.
+  estimator_initial_checkpoint_name = 'model.ckpt-1'
+  save_file = os.path.join(working_dir,
+                           estimator_initial_checkpoint_name)
+  sess = tf.Session(graph=tf.Graph())
+  with sess.graph.as_default():
+    input_features, labels = get_inference_input()
+    dualnet_model.model_fn(
+        input_features, labels, tf.estimator.ModeKeys.PREDICT, params)
+    sess.run(tf.global_variables_initializer())
+    tf.train.Saver().save(sess, save_file)
+
+
+def export_model(working_dir, model_path):
+  """Take the latest checkpoint and export it to model_path for selfplay.
+
+  Assumes that all relevant model files are prefixed by the same name.
+  (For example, foo.index, foo.meta and foo.data-00000-of-00001).
+
+  Args:
+    working_dir: The directory where tf.estimator keeps its checkpoints.
+    model_path: Either a local path or a gs:// path to export model to.
+  """
+  estimator = tf.estimator.Estimator(
+      dualnet_model.model_fn, model_dir=working_dir, params='ignored')
+  latest_checkpoint = estimator.latest_checkpoint()
+  all_checkpoint_files = tf.gfile.Glob(latest_checkpoint + '*')
+  for filename in all_checkpoint_files:
+    suffix = filename.partition(latest_checkpoint)[2]
+    destination_path = model_path + suffix
+    tf.gfile.Copy(filename, destination_path)
+
+
+def train(working_dir, tf_records, generation_num, params):
+  """Train the model for a specific generation.
+
+  Args:
+    working_dir: The model working directory to save model parameters,
+      drop logs, checkpoints, and so on.
+    tf_records: A list of tf_record filenames for training input.
+    generation_num: The generation to be trained.
+    params: hyperparams of the model.
+  """
+  assert generation_num > 0, 'Model 0 is random weights'
```

Comment:
Better to raise a ValueError rather than assert here, since it is the error of the caller.